keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,550 @@
+| kafka    | 2.2.0    |
+| kafka    | 1.2.0    |
+| kafka    | 2.2.1    |
+| kafka    | 1.2.1    |
+
+**Install latest version of the Kafka operator**
+
+In the composite package version strategy we will need to specify how app-version is ordered. 
+```
+kudo install kafka
+```
+
+**for composite version:** 
+
+Installs `kafka-3.1.1-1.0.0`.
+
+**for single version:** 
+
+Installs `kafka-2.2.1`.
+
+**Install the latest Kafka operator with a specific Kafka version**
+
+**for composite version:**
+
+```
+kudo install kafka --app-version 2.3.0
+```
+
+Installs `kafka-2.3.0-1.1.1`.
+
+```
+kudo install kafka --app-version 3.0.0
+```
+
+Installs `kafka-3.0.0-1.1.0`.
+
+**for single version:**
+
+```
+kudo install kafka --app-version 2.3.0
+```
+
+Installs `kafka-1.1.0`.
+
+```
+kudo install kafka --app-version 3.0.0
+```
+
+Installs `kafka-2.2.1`.
+
+**Install a Kafka operator with a specific version**
+
+Let's say a user wants to use Apache Kafka `3.0.0` and even though there's a `3.0.0` with a `1.1.0` revision released, they want the operator with revision `1.0.0` due to reasons like:
+- revision `1.1.0` introduced a bug for which there's still no released fixes
+- revision `1.1.0` changed in a way that will require them to invest time and resources to adapt
+
+**for composite version:**
+
+```
+kudo install kafka --version 3.0.0-1.0.0
+```
+
+**for single version:**
+
+it should be enough with single version
+```
+kudo install kafka --version 2.2.0
+```
+
+But users can also specify the `app-version` flag
+```
+kudo install kafka --version 2.2.0  --app-version 3.0.0
+```
+
+**Install a Kafka operator providing a partial version**
+
+Another interesting thing that having the concept of _version_ contain the _app version_ concatenated with the _operator revision_ is that it would be possible to not have operator users necessarily have to know about the existence of _app version_ and _operator revision_. With regards to versioning, CLI interactions could optionally "flatten" the usage of both flags into just the `--version` flag, which could also be provided as a prefix to be matched.
+
+**for composite version:**
+
+```
+kudo install kafka --version 3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 3.x.x.
+
+```
+kudo install kafka --version 2.3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.x.
+
+```
+kudo install kafka --version 2.3.1
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.1.
+
+**for single version:**
+
+```
+kudo install kafka --app-version 3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 3.x.x.
+
+```
+kudo install kafka --app-version 2.3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.x.
+
+```
+kudo install kafka --app-version 2.3.1
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.1.
+
+**Upgrade Kafka operator**
+
+The `kudo upgrade` commands would look similar as the `kudo install` commands above.
```

Comment:
In case of an upgrade, the user might need to specify if he wants to stay in the same appVersion or want an appVersion upgrade. i.e.
`kudo upgrade` should imho only upgrade from kafka_2.3.1_1.0.0 to kafka_2.3.1_1.1.0, and not to kafka_3.0.0_1.1.0